### PR TITLE
Bugfix for #2165

### LINKED ISSUE #2165: OS X apps table crashes the shell when trying to read restricted Info.plist
Steps to reproduce:
1. Find a `*.app` and `chmod 000 that.app`
2. `osqueryi -A apps`

Trace:
```
 thread #1: tid = 0x8393e, 0x00007fff88b7ff06 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff88b7ff06 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff936f54ec libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff945d56e7 libsystem_c.dylib`abort + 129
    frame #3: 0x00007fff87896f81 libc++abi.dylib`abort_message + 257
    frame #4: 0x00007fff878bca2f libc++abi.dylib`default_terminate_handler() + 243
    frame #5: 0x00007fff89a156c3 libobjc.A.dylib`_objc_terminate() + 124
    frame #6: 0x00007fff878ba19e libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #7: 0x00007fff878b9c12 libc++abi.dylib`__cxa_throw + 121
    frame #8: 0x0000000100b63ae7 osqueryi`boost::filesystem::detail::status(p=<unavailable>, ec=<unavailable>) + 487 at operations.cpp:1658
    frame #9: 0x000000010031002e osqueryi`boost::filesystem::exists(p=0x00007fff5fbf9570) + 46 at operations.hpp:446
    frame #10: 0x000000010030d511 osqueryi`osquery::PlatformFile::PlatformFile(this=0x0000000101f481b0, path="/Users/reed/git/github/darwindumper/Resources/MacGap.app/Contents/Info.plist", mode=41, perms=-1) + 865 at fileops.cpp:78
    frame #11: 0x000000010030d717 osqueryi`osquery::PlatformFile::PlatformFile(this=0x0000000101f481b0, path="/Users/reed/git/github/darwindumper/Resources/MacGap.app/Contents/Info.plist", mode=41, perms=-1) + 55 at fileops.cpp:30
    frame #12: 0x0000000100308d42 osqueryi`osquery::OpenReadableFile::OpenReadableFile(this=0x00007fff5fbf9eb8, path=0x00007fff5fbfab28) + 866 at filesystem.cpp:85
    frame #13: 0x0000000100307d1b osqueryi`osquery::OpenReadableFile::OpenReadableFile(this=0x00007fff5fbf9eb8, path=0x00007fff5fbfab28) + 43 at filesystem.cpp:79
    frame #14: 0x00000001002ff0f0 osqueryi`osquery::readFile(path=0x00007fff5fbfab28, size=0, block_size=4096, dry_run=true, preserve_time=false, predicate=function<void (std::__1::basic_string<char> &, unsigned long)> @ 0x00007fff5fbf9f90)>) + 128 at filesystem.cpp:109
    frame #15: 0x0000000100300129 osqueryi`osquery::readFile(path=0x00007fff5fbfab28, content="", size=0, dry_run=true, preserve_time=false) + 233 at filesystem.cpp:171
    frame #16: 0x00000001003002dd osqueryi`osquery::readFile(path=0x00007fff5fbfab28) + 237 at filesystem.cpp:187
    frame #17: 0x00000001002fa22c osqueryi`osquery::parsePlist(path=0x00007fff5fbfab28, tree=0x00007fff5fbfab70) + 140 at plist.mm:169
    frame #18: 0x00000001005cd4e0 osqueryi`osquery::tables::genApps(context=0x00007fff5fbfc8a8) + 4848 at apps.mm:273
    frame #19: 0x000000010085ff40 osqueryi`osquery::appsTablePlugin::generate(this=0x000000010183f720, request=0x00007fff5fbfc8a8) + 160 at additional_amalgamation.cpp:243
    frame #20: 0x000000010032ca48 osqueryi`osquery::RegistryFactory::callTable(table_name="apps", context=0x00007fff5fbfc8a8, response=size=0) + 840 at registry.cpp:434
    frame #21: 0x00000001005bb532 osqueryi`osquery::tables::sqlite::xFilter(pVtabCursor=0x0000000101a02320, idxNum=0, idxStr=0x0000000000000000, argc=0, argv=0x0000000101a01280) + 11858 at virtual_table.cpp:454
    frame #22: 0x0000000100087194 osqueryi`sqlite3VdbeExec(p=0x0000000101a00a10) + 37716 at sqlite3.c:83264
    frame #23: 0x00000001000479e1 osqueryi`sqlite3Step(p=0x0000000101a00a10) + 513 at sqlite3.c:75177
    frame #24: 0x00000001000475ef osqueryi`sqlite3_step(pStmt=0x0000000101a00a10) + 143 at sqlite3.c:75238
    frame #25: 0x0000000100009cae osqueryi`shell_exec(zSql="SELECT * FROM apps", xCallback=(osqueryi`shell_callback(void*, int, char**, char**, int*) at shell.cpp:437), pArg=0x00007fff5fbfdce0, pzErrMsg=0x0000000000000000)(void*, int, char**, char**, int*), callback_data*, char**) + 734 at shell.cpp:751
    frame #26: 0x00000001000075f9 osqueryi`do_meta_command(zLine=".all", p=0x00007fff5fbfee58) + 1817 at shell.cpp:1093
    frame #27: 0x0000000100006685 osqueryi`osquery::launchIntoShell(argc=1, argv=0x00007fff5fbff7f0) + 3445 at shell.cpp:1482
    frame #28: 0x00000001000395d8 osqueryi`main(argc=1, argv=0x00007fff5fbff7f0) + 984 at shell.cpp:105
    frame #29: 0x00007fff98e3a5ad libdyld.dylib`start + 1
```

This only occurs in the shell, the daemon has an exception wrapper around executing queries and will report a log line: "failed to execute ..." if this occurs. See the `--registry_exceptions` configuration flag that is default `false`.

### PR DESCRIPTION
**Pull request by yying**
This PR is intended to fix issues in #2165

Closes #2165

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/filesystem/windows/fileops.cpp [hunk 1]
```diff
@@ -929,6 +933,7 @@ bool platformChmod(const std::string& path, mode_t perms) {
 
 std::vector<std::string> platformGlob(const std::string& find_path) {
   fs::path full_path(find_path);
+  boost::system::error_code ec;
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: Push this into the loop context, replacing line 989 works with the object construction works.

File: osquery/filesystem/windows/fileops.cpp [hunk 2]
```diff
@@ -1003,7 +1009,9 @@ std::vector<std::string> platformGlob(const std::string& find_path) {
       for (auto& result : wf.get()) {
         if (std::regex_match(result.filename().string(), component_pattern)) {
           auto result_path = result.make_preferred().string();
-          if (fs::is_directory(result)) {
+          ec.clear();
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: Construct here.

File: osquery/filesystem/windows/fileops.cpp [hunk 3]
```diff
@@ -1013,7 +1021,9 @@ std::vector<std::string> platformGlob(const std::string& find_path) {
       WindowsFindFiles wf(valid_path / full_path.filename());
       for (auto& result : wf.get()) {
         auto result_path = result.make_preferred().string();
-        if (fs::is_directory(result)) {
+        ec.clear();
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: And here is fine.

**Comment by theopolis** [comment]
ok to test

File: osquery/filesystem/windows/fileops.cpp [hunk 4]
```diff
@@ -1080,8 +1089,9 @@ static std::string normalizeDirPath(const fs::path &path) {
   // NTFS is case insensitive, to normalize, make everything uppercase
   ::CharUpperA(&final_path[0]);
 
+  boost::system::error_code ec;
   std::string normalized_path(&final_path[0], nret);
-  if (fs::is_directory(normalized_path) && normalized_path[nret - 1] != '\\') {
+  if ((fs::is_directory(normalized_path, ec) && ec.value() == errc::success) && normalized_path[nret - 1] != '\\') {
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: Still no `make format` on Windows? :(

File: osquery/filesystem/windows/fileops.cpp [hunk 5]
```diff
@@ -1029,7 +1038,7 @@ boost::optional<std::string> getHomeDirectory() {
   auto value = getEnvVar("USERPROFILE");
   if (value.is_initialized()) {
     return value;
-  } else if (SUCCEEDED(::SHGetFolderPathA(NULL, CSIDL_PROFILE, NULL, 0, &profile[0]))) {
+  } else if (SUCCEEDED(::SHGetFolderPathA(nullptr, CSIDL_PROFILE, nullptr, 0, &profile[0]))) {
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: Looks like +80 characters.

File: osquery/filesystem/posix/fileops.cpp [hunk 6]
```diff
@@ -75,7 +76,8 @@ PlatformFile::PlatformFile(const std::string& path, int mode, int perms) {
     perms = 0666;
   }
 
-  if (check_existence && !fs::exists(path.c_str())) {
+  boost::system::error_code ec;
+  if (check_existence && (!fs::exists(path.c_str(), ec) || ec.value() != errc::success)) {
```
Comment (theopolis) [inline osquery/filesystem/posix/fileops.cpp]: nit, +80 chars

**Comment by osqueryer** [comment]
Build origin/pr/2167/merge (Job: 2760) failed :-1:

**Comment by osqueryer** [comment]
Build origin/pr/2167/merge (Job: 2763) is good! :+1:

**Pull request merged**